maltoolbox/attackgraph/attackgraph.py

```diff
@@ -493,6 +333,176 @@ def attach_attackers(self) -> None:
+                        'expression chain is missing sub link.')
+                accumulated_target_assets = list(target_assets)
+                new_target_assets = list(target_assets)
+                while new_target_assets:
+                    additional_assets = self._follow_expr_chain(
+                        model, new_target_assets, expr_chain.sub_link)
+
+                    new_target_assets = []
+                    for additional_asset in additional_assets:
+                        if additional_asset not in accumulated_target_assets:
+                            accumulated_target_assets.append(additional_asset)
+                            new_target_assets.append(additional_asset)
+
+                return accumulated_target_assets
+
+            case 'subType':
+                if not expr_chain.sub_link:
+                    raise LanguageGraphException('"subType" step '
+                        'expression chain is missing sub link.')
+                new_target_assets = []
+                for target_asset in target_assets:
+                    assets = self._follow_expr_chain(
+                        model, target_assets,
+                        expr_chain.sub_link)
+                    new_target_assets.extend(assets)
+
+                selected_new_target_assets = []
+                for asset in new_target_assets:
+                    lang_graph_asset = self.lang_graph.assets[asset.type]
+                    if not lang_graph_asset:
+                        raise LookupError(
+                            f'Failed to find asset \"{asset.type}\" in the '
+                            'language graph.'
+                        )
+                    lang_graph_subtype_asset = expr_chain.subtype
+                    if not lang_graph_subtype_asset:
+                        raise LookupError(
+                            'Failed to find asset "%s" in the '
+                            'language graph.' % expr_chain.subtype
+                        )
+                    if lang_graph_asset.is_subasset_of(
+                            lang_graph_subtype_asset):
+                        selected_new_target_assets.append(asset)
+
+                return selected_new_target_assets
+
+            case 'collect':
+                if not expr_chain.left_link:
+                    raise LanguageGraphException('"collect" step expression chain'
+                        ' is missing the left link.')
+                if not expr_chain.right_link:
+                    raise LanguageGraphException('"collect" step expression chain'
+                        ' is missing the right link.')
+                lh_targets = self._follow_expr_chain(
+                    model,
+                    target_assets,
+                    expr_chain.left_link
+                )
+                rh_targets = self._follow_expr_chain(
+                    model,
+                    lh_targets,
+                    expr_chain.right_link
```

Comment:
Will this be sth else than None? If yes, I have not understood what is going on, no need to explain. If no, maybe have it explicitly say None?

```suggestion
                    None,
```

Reply:
This is not at all a criticism, but it should be blatantly obvious by now that you can't ask my brain to not explain something. It's just not how it works.

Your overall gut feeling is correct, but it's what led to my initial wrong assumption. Yes, most often the right side of a collect will be None. However, you can have stuff like (exprA).(exprB), exprB would have to contain the attack step somewhere in there. So that resulting expression would have exprA in the left link and exprB in the right link, and exprB has to contain at least one other collect expression that has the attack step on the right hand side which would be a None right link.

Did I "not explain" that clearly enough?

Reply:
Ah right, I see it now. Thanks for clarifying !

Reply:
Just for future reference in case I/we need it. You don't actually need the brackets for exprA, any expression that has exprA.(exprB) has a non-None right link because the right link will always be a collect that eventually has to contain a None on a right link.